stdlib/string.gr

```diff
@@ -872,6 +872,156 @@ export let endsWith = (search: String, string: String) => {
   }
 }
 
+/**
+ * Replaces the first appearance of the search pattern in the string with the replacement value.
+ *
+ * @param searchPattern: The string to replace
+ * @param replacement: The string to replace with
+ * @param string: The string to replace on
+ * @returns A new string with the search pattern replaced
+ *
+ * @example String.replaceFirst("Hello", "Hi", "Hello World Hello") == "Hi World Hello"
+ *
+ * @since v0.5.3
+ */
+@unsafe
+export let replaceFirst =
+  (
+    searchPattern: String,
+    replacement: String,
+    string: String,
+  ) => {
+  let (+) = WasmI32.add
+  let (-) = WasmI32.sub
+  let (>) = WasmI32.gtU
+  let (<) = WasmI32.ltU
+  let (==) = WasmI32.eq
+
+  let mut patternPtr = WasmI32.fromGrain(searchPattern)
+  let mut stringPtr = WasmI32.fromGrain(string)
+  let mut replacementPtr = WasmI32.fromGrain(replacement)
+
+  let patternLen = WasmI32.load(patternPtr, 4n)
+  let stringLen = WasmI32.load(stringPtr, 4n)
+  let replacementLen = WasmI32.load(replacementPtr, 4n)
+
+  patternPtr += 8n
+  stringPtr += 8n
+  replacementPtr += 8n
+  // Bail if search str is longer then the string
+  if (stringLen < patternLen) {
+    string
+  } else {
+    let str = allocateString(stringLen - patternLen + replacementLen)
```

Comment:
This will leak memory because we never free it in the case we get no match. But instead of doing that, you can move all of the logic to create the new string to the `!found` else branch, so we only do it if we found a position we need to replace. (I'd swap the logic though, so it'd be `if (found) <allocate new string, do copies, return string> else <return original string>`.)